packages/tre/src/plugins/r2/gateway.ts

```diff
@@ -177,22 +359,298 @@ function sqlStmts(db: TypedDatabase) {
+      });
+
+      return oldBlobIds;
+    }),
   };
 }
 
 export class R2Gateway {
+  // Minimum multipart part upload size is configurable, so we can use smaller
+  // values in tests
+  /** @internal */
+  static _MIN_MULTIPART_PART_SIZE = 5 * 1024 * 1024;
+
   readonly #storage: NewStorage;
   readonly #stmts: ReturnType<typeof sqlStmts>;
 
-  constructor(private readonly log: Log, legacyStorage: Storage) {
+  // Multipart uploads are stored as multiple blobs. Therefore, when reading a
+  // multipart upload, we'll be reading multiple blobs. When an object is
+  // deleted, all its blobs are deleted in the background.
+  //
+  // Normally for single part objects, this is fine, since we'd open a handle to
+  // a single blob, which we'd have until we closed it, at which point the blob
+  // may be deleted. With multipart, we don't want to open handles for all blobs
+  // as we could hit open file descriptor limits. Similarly, we don't want to
+  // read all blobs first, as we'd have to buffer them.
+  //
+  // Instead, we set up in-process locking on blobs needed for multipart reads.
+  // When we start a multipart read, we acquire all the blobs we need, then
+  // release them as we've streamed each part. Multiple multipart reads may be
+  // in-progress at any given time, so we use a wait group.
+  //
+  // This assumes we only ever have a single Miniflare instance operating on a
+  // blob store, which is always true for in-memory stores, and usually true for
+  // on-disk ones. If we really wanted to do this properly, we could store the
+  // bookkeeping for the wait group in SQLite, but then we'd have to implement
+  // some inter-process signalling/subscription system.
+  readonly #inUseBlobs = new Map<BlobId, WaitGroup>();
+
+  constructor(
+    private readonly log: Log,
+    legacyStorage: Storage,
+    private readonly timers: Timers
+  ) {
     this.#storage = legacyStorage.getNewStorage();
     this.#storage.db.pragma("case_sensitive_like = TRUE");
     this.#storage.db.exec(SQL_SCHEMA);
     this.#stmts = sqlStmts(this.#storage.db);
   }
 
-  #backgroundDelete(blobId: string) {
-    void this.#storage.blob.delete(blobId).catch(() => {});
+  #acquireBlob(blobId: BlobId) {
+    let waitGroup = this.#inUseBlobs.get(blobId);
+    if (waitGroup === undefined) {
+      waitGroup = new WaitGroup();
+      this.#inUseBlobs.set(blobId, waitGroup);
+      waitGroup.add();
+      // Automatically remove the wait group once this blob is fully released
+      waitGroup.wait().then(() => this.#inUseBlobs.delete(blobId));
+    } else {
+      waitGroup.add();
+    }
+  }
+
+  #releaseBlob(blobId: BlobId) {
+    this.#inUseBlobs.get(blobId)?.done();
+  }
+
+  #backgroundDelete(blobId: BlobId) {
+    this.timers.queueMicrotask(async () => {
+      // Wait for all multipart gets using this blob to complete
+      await this.#inUseBlobs.get(blobId)?.wait();
+      return this.#storage.blob.delete(blobId).catch(() => {});
```

Comment:
I guess there is no chance of a race-condition here where a new client starts using a blob between the wait group being released and the blob being deleted?